proposals/007-plugin-api-to-select-tls-credentials-for-server-connection.md

```diff
@@ -0,0 +1,178 @@
+
+```java
+package io.kroxylicious.proxy.tls;
+
+import java.security.PrivateKey;
+import java.security.cert.Certificate;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+import io.kroxylicious.proxy.authentication.ClientSaslContext;
+
+/**
+ * Implemented by a {@link io.kroxylicious.proxy.filter.Filter} that provides
+ * the credentials for the TLS connection between the proxy and the Kafka server.
+ */
+public interface ServerTlsCredentialSupplier {
+    /**
+     * Return the TlsCredentials for the connection.
+     * @param context The context.
+     * @return the TlsCredentials for the connection.
+     */
+    CompletionStage<TlsCredentials> tlsCredentials(Context context);
+}
+```
+
+Where the `Context` will be a inner interface:
+
+```java
+    /**
+     * The context API for {@link ServerTlsCredentialSupplier}.
+     * This is implemented by the runtime for use by plugins.
+     */
+    interface Context {
+        Optional<ClientTlsContext> clientTlsContext();
+
+        /**
+         * Returns the default credentials for this target cluster (e.g. from the proxy configuration file).
+         * Implementations of {@link ServerTlsCredentialSupplier} may use this as a fall-back
+         * or default, for example if the apply a certificiate-per-client-principal pattern
+         * but are being used with an anonymous principal.
+         * @return the default credentials.
+         */
+        TlsCredentials defaultTlsCredentials();
+
+        /**
+         * <p>Factory methods for creating TLS credentials for the given parameters.</p>
+         *
+         * <p>The equivalent method on {@code FilterFactoryContext} can be used when the credentials
+         * are known at plugin configuration time.</p>
+         *
+         * @param key The key corresponding to the given client certificate.
+         * @param certificateChain The client certificate corresponding to the given {@code key}, plus any intermediate certificates forming the certificate chain up to (but not including) the TLS certificate trusted by the peer.
+         * @return The TLS credentials instance.
+         * see io.kroxylicious.proxy.filter.ServerTlsCredentialSupplierFactory.Context#tlsCredentials(PrivateKey, Certificate[])
+         */
+        TlsCredentials tlsCredentials(Certificate certificate,
+                                      PrivateKey key,
+                                      Certificate[] intermediateCertificates);
+    }
+```
+
+The `Context` is intentionally quite narrow in terms of the information it exposes to plugin implementations, but it is extensible in the future as use cases emerge for exposing more runtime information to implementations.
+
+And `TlsCredentials` looks like this:
+
+```java
+package io.kroxylicious.proxy.authentication;
+
+interface TlsCredentials {
+  /* Intentionally empty: implemented and accessed only in the runtime */
+}
+```
+
+// TODO Why not use the JDK's `X500PrivateCredential`?
```

Comment:
In the end did you choose not to do this because of how in-flux JAAS looked?

Reply:
Although `X500PrivateCredential` has a cert and private key, it doesn't directly model intermediate certificates. Instead is has an `alias` property, which refers to an entry in some `KeyStore` that the instantiator and user of the `X500PrivateCredential` instance have to mutually have a reference to (which is left to the application to sort out).

So using it would involve some costs:
* extra hoop jumping in terms of passing around a KeyStore
* and also any credentials needed to access the entry within the key store. 
 
Meanwhile the only benefit I can see right now is "It's a JDK type". 

So over all I don't think it's worth it.